src/wasm/wasm-type.cpp

```diff
@@ -2276,191 +2271,174 @@ void TypeBuilder::createRecGroup(size_t index, size_t length) {
+    return false;
+  }
+  SubTyper typer;
+  switch (sub.kind) {
+    case HeapTypeInfo::SignatureKind:
+      return typer.isSubType(sub.signature, super.signature);
+    case HeapTypeInfo::StructKind:
+      return typer.isSubType(sub.struct_, super.struct_);
+    case HeapTypeInfo::ArrayKind:
+      return typer.isSubType(sub.array, super.array);
+  }
+  WASM_UNREACHABLE("unknown kind");
+}
+
+void updateReferencedHeapTypes(
+  std::unique_ptr<HeapTypeInfo>& info,
+  const std::unordered_map<HeapType, HeapType>& canonicalized) {
+  // Update the reference types that refer to canonicalized heap types to be
+  // their canonical versions. Update the Types rather than the HeapTypes so
+  // that the validation of supertypes sees canonical types.
+  struct ChildUpdater : TypeGraphWalkerBase<ChildUpdater> {
+    const std::unordered_map<HeapType, HeapType>& canonicalized;
+    bool isTopLevel = true;
+
+    ChildUpdater(const std::unordered_map<HeapType, HeapType>& canonicalized)
+      : canonicalized(canonicalized) {}
+
+    void scanType(Type* type) {
+      isTopLevel = false;
+      if (type->isRef()) {
+        auto ht = type->getHeapType();
+        if (auto it = canonicalized.find(ht); it != canonicalized.end()) {
+          *type = Type(it->second, type->getNullability());
+        } else if (isTemp(*type) && !isTemp(ht)) {
+          *type = Type(ht, type->getNullability());
+        }
+      } else if (type->isTuple()) {
+        TypeGraphWalkerBase<ChildUpdater>::scanType(type);
+      }
     }
-    HeapType getAsHeapType() {
-      if (auto* type = getHeapType()) {
-        return *type;
+
+    void scanHeapType(HeapType* type) {
+      if (isTopLevel) {
+        TypeGraphWalkerBase<ChildUpdater>::scanHeapType(type);
       }
-      return asHeapType(*getHeapTypeInfo());
     }
   };
 
-  using ReplacementMap = std::unordered_map<HeapType, Replacement>;
-
-  // Updates the `results` and `newInfo` lists, but does not modify any of the
-  // infos to update HeapType use sites.
-  void updateShallow(ReplacementMap& replacements);
-  // Updates the HeapType use sites within `info`.
-  void updateUses(ReplacementMap& replacements,
-                  std::unique_ptr<HeapTypeInfo>& info);
+  // Update the children.
+  ChildUpdater updater(canonicalized);
+  auto root = asHeapType(info);
+  updater.walkRoot(&root);
 
-#if TRACE_CANONICALIZATION
-  void dump() {
-    IndexedTypeNameGenerator print(results);
-    std::cerr << "Results:\n";
-    for (size_t i = 0; i < results.size(); ++i) {
-      std::cerr << i << ": " << print(results[i]) << "\n";
-    }
-    std::cerr << "NewInfos:\n";
-    for (size_t i = 0; i < newInfos.size(); ++i) {
-      std::cerr << print(asHeapType(newInfos[i])) << "\n";
+  // Update the supertype.
+  if (info->supertype) {
+    HeapType super(uintptr_t(info->supertype));
+    if (auto it = canonicalized.find(super); it != canonicalized.end()) {
+      info->supertype = getHeapTypeInfo(it->second);
     }
```

Comment:
The supertype is a HeapType rather than a reference Type, so we only need to update it if the current HeapType is known to be non-canonical, i.e. is in the `canonicalized` map. In contrast, the code above needs to replace all non-canonical reference types, either because they point to a non-canonical type (which we detect by looking in `canonicalized`) or because they are a temporary, non-canonical type owned by the TypeBuilder even though they point to a canonical heap type (which we detect with `isTemp`). The latter case cannot happen for the supertype.